File: src/arrays/matrices/python/mmult.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Algorithms and Programming II                                April 19, 2022
IST 2089
Prof. M Diaz-Maldondo

Synopsis:
Solves the following matrix multiplication problem:

                            D = B' * A * C

and outputs the requested element on the console.


Copyright (c) 2022 Misael Diaz-Maldonado
This file is released under the GNU General Public License as published
by the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
"""


from numpy import array
from numpy import matmul


A = array([
             9, 45, 39, 16, 33, 58,
            60, 54, 11,  3, 24, 47,
            56, 23, 63,  7, 18, 53,
            22, 17, 12, 38, 30, 21,
            57, 20, 35, 36, 19, 10,
            14, 32,  6, 28,  8, 51
]).reshape([6, 6])


B = array([
            21, 25, 17, 33, 54,  3,
            48, 40,  8, 63,  9, 53,
            13, 11, 24,  5, 60, 52,
             1, 10, 56,  7, 55, 44,
            26, 22, 28, 62, 45, 41,
            39, 18, 31, 29,  4, 50
]).reshape([6, 6])


C = array([
            30, 60, 36, 14, 61, 34,
            57, 18, 43, 13, 11, 53,
            20,  5, 23, 26, 27, 49,
            37, 28, 62, 10,  1,  7,
            46, 21, 22, 58, 52,  9,
            16, 33, 25, 45, 12, 40
]).reshape([6, 6])


D = matmul(matmul(B.T, A), C)


"""
Answer:
Input D(1, 2) (integer) in the answer field."
"""
i, j = 1, 2
ans = f"D({i}, {j}) = {D[i-1][j-1]}"
print(ans)

File: src/arrays/matrices/matlab/mmult.m
% Algorithms and Programming II                              April 19, 2022
% IST 2089
% Prof. M. Diaz-Maldonado
%
% Synopsis:
% Code shows how to perform some matrix algebra operations in MATLAB.
%
%
% Copyright (c) 2022 Misael Diaz-Maldonado
% This file is released under the GNU General Public License as published
% by the Free Software Foundation, either version 3 of the License, or
% (at your option) any later version.
%
%
% References:
% [0] A Gilat, MATLAB: An Introduction with Applications, 6th edition.

clear
close all
clc

% initializes the 5 x 5 matrices A, B, and C
A = [52, 41, 48, 24, 50
     54, 29, 40, 13, 11
     25, 55, 43,  5, 51
     12, 49, 53, 31,  9
     56,  8, 35, 57, 19];

B = [45,  1, 48, 49, 61
     40, 55, 33, 34,  6
      4, 13, 27, 46, 31
     26, 56, 16, 62, 42
     51, 19, 24, 23, 50];

C = [45, 51,  1, 42, 21
     47, 28,  6,  5, 15
     20, 37, 40, 36, 24
     53, 44,  4, 17, 11
     52, 33, 34,  9, 58];

% obtains the trace of the resulting matrix
tr = trace(B * A * C');
fprintf("trace: %d\n", tr);
